include/oneapi/dpl/internal/distributed_ranges_impl/sp/init.hpp

```diff
@@ -0,0 +1,165 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DR_SP_INIT_HPP
+#define _ONEDPL_DR_SP_INIT_HPP
+
+#include <cassert>
+#include <memory>
+#include <span>
+#include <sycl/sycl.hpp>
+#include <type_traits>
+#include <vector>
+
+#include "oneapi/dpl/execution"
+
+#include "algorithms/execution_policy.hpp"
+#include "util.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+namespace __detail
+{
+
+inline sycl::context* global_context_;
+
+inline std::vector<sycl::device> devices_;
+
+inline std::vector<sycl::queue> queues_;
+
+inline std::vector<oneapi::dpl::execution::device_policy<>> dpl_policies_;
+
+inline std::size_t ngpus_;
+
+inline sycl::context&
+global_context()
+{
+    return *global_context_;
+}
+
+inline std::size_t
+ngpus()
+{
+    return ngpus_;
+}
+
+inline std::span<sycl::device>
+global_devices()
+{
+    return devices_;
+}
+
+} // namespace __detail
+
+inline sycl::context&
+context()
+{
+    return __detail::global_context();
+}
+
+inline std::span<sycl::device>
+devices()
+{
+    return __detail::global_devices();
+}
+
+inline std::size_t
+nprocs()
+{
+    return __detail::ngpus();
+}
+
+inline sycl_device_collection par_unseq;
+
+template <stdrng::range R>
+inline void
+init(R&& devices) requires(std::is_same_v<sycl::device, std::remove_cvref_t<stdrng::range_value_t<R>>>)
+{
+    __detail::devices_.assign(stdrng::begin(devices), stdrng::end(devices));
+    __detail::global_context_ = new sycl::context(__detail::devices_);
```

Comment:
From what I can tell, there is nothing preventing a user from calling this `init` function multiple times, which would result in a memory leak here. Would it help if the context was stored in a `shared_ptr` or `unique_ptr`? Or is there something else that I am missing?